### PR TITLE
Add lowering for adaptive_max_pool2d

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120460
* __->__ #120254

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4138,6 +4137,124 @@ def fn(idx):
...
+                else:
+                    maxindex = ops.where(ops.gt(val, maxval), index, maxindex)
+            if maxval is None:
+                maxval = val
+            else:
+                maxval = ops.maximum(val, maxval)
+            if return_index:
+                return maxindex
+            else:
+                return maxval
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4185]: Indentation error? You're returning in the first iteration every time which is why you only have one load in the resulting kernel.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -4138,6 +4137,124 @@ def fn(idx):
     # TODO: should we force these to be realized?
     return rv
 
+def _adaptive_pooling_idx_max(kernel_maxes, in_sizes, out_sizes, return_index, loader):
```
Comment (lezcano) [inline torch/_inductor/lowering.py:4140]: Did you end up implementing this copying `adaptive_pooling_avg`? If so, can you leave a note saying we should deduplicate this code once https://github.com/pytorch/pytorch/pull/116418 is merged.

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (amjames) [reply torch/_inductor/lowering.py:4185]: 🤦 yep, that did the trick. Thanks

**Comment by amjames** [comment]
Timing data below. Note: To avoid the fallback to eager condition the output_size is varied as a fraction of the input size (0.25, 0.5. 0.75). 



```
[-------------------------- adaptive_max_pool2d-cuda --------------------------]
                                                            |  Compile  |  Eager
32 threads: --------------------------------------------------------------------
      (512, 3, 100, 100), adaptive_max_pool2d((25, 25))     |     254   |    264
      (512, 3, 100, 100), adaptive_max_pool2d((50, 50))     |     368   |    429
      (512, 3, 100, 100), adaptive_max_pool2d((75, 75))     |     599   |    853
      (512, 3, 100, 300), adaptive_max_pool2d((25, 75))     |     783   |    824
      (512, 3, 100, 300), adaptive_max_pool2d((50, 150))    |    1131   |   1485
      (512, 3, 100, 300), adaptive_max_pool2d((75, 225))    |    1770   |   3070
      (512, 3, 300, 100), adaptive_max_pool2d((75, 25))     |     785   |    772
      (512, 3, 300, 100), adaptive_max_pool2d((150, 50))    |    1101   |   1299
      (512, 3, 300, 100), adaptive_max_pool2d((225, 75))    |    1805   |   2690
      (512, 3, 300, 300), adaptive_max_pool2d((75, 75))     |    2284   |   2507
      (512, 3, 300, 300), adaptive_max_pool2d((150, 150))   |    3323   |   4540
      (512, 3, 300, 300), adaptive_max_pool2d((225, 225))   |    5330   |   9029
      (1024, 3, 100, 100), adaptive_max_pool2d((25, 25))    |     504   |    522
      (1024, 3, 100, 100), adaptive_max_pool2d((50, 50))    |     733   |    846
      (1024, 3, 100, 100), adaptive_max_pool2d((75, 75))    |    1193   |   1650
      (1024, 3, 100, 300), adaptive_max_pool2d((25, 75))    |    1503   |   1635
      (1024, 3, 100, 300), adaptive_max_pool2d((50, 150))   |    2199   |   2872
      (1024, 3, 100, 300), adaptive_max_pool2d((75, 225))   |    3536   |   6071
      (1024, 3, 300, 100), adaptive_max_pool2d((75, 25))    |    1506   |   1537
      (1024, 3, 300, 100), adaptive_max_pool2d((150, 50))   |    2199   |   2600
      (1024, 3, 300, 100), adaptive_max_pool2d((225, 75))   |    3556   |   5384
      (1024, 3, 300, 300), adaptive_max_pool2d((75, 75))    |    4506   |   4990
      (1024, 3, 300, 300), adaptive_max_pool2d((150, 150))  |    6591   |   9100
      (1024, 3, 300, 300), adaptive_max_pool2d((225, 225))  |   10590   |  18050

```

**Comment by amjames** [comment]
@lezcano yes, one kernel.

**Review by lezcano — COMMENTED** [review]
SG. CI failures are real tho.

**Review by lezcano — APPROVED** [review]
Nice.

cc @isuruf. We should clean all these once all your work on adaptive_avg_* is merged.

**Pull request closed without merging**